### ISSUE TITLE
Document the built-in rules' emphasis on standard syntax better

### ISSUE THREAD

**Issue by certainlyakey** [stylelint/stylelint#2979]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

If a complex selector contains an interpolated SASS variable stylelint does not see a type selector inside it and doesn't report it.

> Which rule, if any, is this issue related to?

`selector-max-type`

> What CSS is needed to reproduce this issue?

```css
#{$var} span {
    width:inherit;
}
```

where $var can be anything (class, another type selector...). In the last case, the type selector contained in the variable is not reported as well.

> What stylelint configuration is needed to reproduce this issue?
```json
{
  "rules": {
    "selector-max-type": 0
  }
}
```

> Which version of stylelint are you using?

`8.2.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

`gulp-stylelint`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to SASS variable interpolation.

> What did you expect to happen?

An error should be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

Nothing was reported. If you make the type selector to be a part of selector that does not contain a variable (or just use it alone) it does report.

**Comment by hudochenkov** [comment]
Thank you for your report!

This rule works like expected. It ignores non-standard selector because it's not possible to be sure all the time that we can correctly identify a problem in non-standard selectors.

**Comment by certainlyakey** [comment]
Hmm isn't stylelint supposed to work with preprocessors as well?
From the README:
> Supports CSS-like syntaxes

**Comment by hudochenkov** [comment]
It does, but some parts deliberately ignored. Linters don't compile preprocessors code, but they try to analyze as much as possible. And it's not always possible to analyze everything without compiling.

**Comment by jeddy3** [comment]
@certainlyakey Thanks for the report and for using the template.

> It probably should be said in the readme 

I'm going to reopen this as a documentation issue. As a rule of thumb, I think that if we can't cite from the docs when answering a question or issue, then we probably need to add to the docs.

> some sort of documentation what parts of preprocessors' functionality are deliberately ignored.

@certainlyakey Dealing with non-standard syntax is difficult and it's something we've been battling with since the early days of stylelint. As there is an [ever growing list](https://github.com/postcss/postcss/blob/master/docs/plugins.md) of processor features, I don't think it's feasible to list everything that is ignored. We touch on this in the [CSS processors doc](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md). However, I think we can try to make things clearer.

> For example, a very recent issue [here](https://github.com/stylelint/stylelint/issues/2946) deals with SASS maps which are a purely preprocessor feature as well.

Well spotted. All the rules (with the exception of `indentation`) use a series of utils ([`isStandardSyntax*`](https://github.com/stylelint/stylelint/tree/master/lib/utils)) that _try_ to whitelist standard CSS syntax. Again it's difficult as the landscape is always shifting, but this whitelisting approach has, so far, proven invaluable at keeping the scope of the rules manageable. We made a pragmatic exception when it came to the `indentation` rule due to the complexity of the rule and the fact that it touches every part of the stylesheet. We can better document this.

I think the best way to resolve this is to emphasise that all the built-in rules (with the exception of indentation) will ignore [_things_](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/about-rules.md) that contain non-standard syntax e.g. variable interpolation. I think we can do this in the main README and above [the list of rules](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules.md).

> From the README:
>> Supports CSS-like syntaxes

Funnily enough, we just changed that from "Understands CSS-like syntaxes" [just a few days ago](https://github.com/stylelint/stylelint/pull/2968) as we thought "Understands" was too ambiguous. Now I see that "Supports" is no better. I'll try to address that too. We should make clear that, even though the _built-in rules_ are geared towards standard CSS syntax, stylelint:

1. Can parse non-standard syntax.
2. Is extensible and the community can (and is) writing rules for non-standard syntax (e.g. [`stylelint-scss`](https://github.com/kristerkari/stylelint-scss)

**Comment by certainlyakey** [comment]
OK, thanks for the clarification and the really clear write-up! 
The project could surely make an advantage of documenting better that preprocessor related features are ignored by default.
> Is extensible and the community can (and is) writing rules for non-standard syntax (e.g. stylelint-scss)
Would it be a good idea to post this as a (pull-)request to the `stylelint-scss` plugin?

**Comment by hudochenkov** [comment]
Done in https://github.com/stylelint/stylelint/pull/2981.
